pkg/policy/rule.go

```diff
@@ -371,28 +421,40 @@ func (r *rule) resolveL4Policy(ctx *SearchContext, state *traceState, result *L4
 	state.selectRule(ctx, r)
 	found := 0
 
-	if !ctx.EgressL4Only {
-		if len(r.Ingress) == 0 {
-			ctx.PolicyTrace("    No L4 rules\n")
+	// Always resolve both ingress and egress, in order to always generate the same
+	// L4Policy. This is required for resolving visibility rules. But disable tracing
+	// for the rules that were not requested for tracing.
+	ingressCtx := *ctx
```

Comment:
Be aware you are *not* making a deep copy of `From` and `To`, can you guarantee that `mergeL4` will never modify the `From` nor `To`?

Reply:
I think it'd be nice to add some guarantee of this, since I'd expect that the SearchContext should be purely an input to the resolution to filter down which policies are relevant.

Reply:
We never modify the pointed variables, only fields directly in SearchContext.

Reply:
But this will be a problem in the future, I (or any external developer) might not know that we never modify the pointed variables. IMO it doesn't make sense to only be able to change some parts of a struct but not the others.

Reply:
I agree that this ugly. And there are currently several places in the policy package where the SearchContext is modified. I'm planning a redesign of the policy code. Tracing needs to be fixed anyway so we report which _set of rules_ denies or allows a flow, in order to troubleshoot policies. That will require invasive changes.

Reply:
I will work on that in the coming release. It's just too big for this particular PR.